internal/provider/resource_secret_access.go

```diff
@@ -0,0 +1,332 @@
+		}
+	}
+
+	// grant access to applications that are in the plan but not in the state
+	if applicationsInPlanNotInState != nil {
+		err = s.client.Secrets.GrantSecretAccess(&juju.GrantRevokeSecretAccessInput{
+			ModelName:    state.Model.ValueString(),
+			SecretId:     state.SecretId.ValueString(),
+			Applications: applicationsInPlanNotInState.([]string),
+		})
+		if err != nil {
+			resp.Diagnostics.AddError("Failed to grant secret access", err.Error())
+			return
+		}
+	}
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("update secret access %q", state.SecretId))
+}
+
+func (s *secretAccessResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// Prevent panic if the provider has not been configured.
+	if s.client == nil {
+		addClientNotConfiguredError(&resp.Diagnostics, "secret_access", "delete")
+		return
+	}
+
+	var state secretAccessResourceModel
+
+	// Read Terraform configuration state into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	applications := make([]string, len(state.Applications.Elements()))
+	resp.Diagnostics.Append(state.Applications.ElementsAs(ctx, &applications, false)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	err := s.client.Secrets.RevokeSecretAccess(&juju.GrantRevokeSecretAccessInput{
+		ModelName:    state.Model.ValueString(),
+		SecretId:     state.SecretId.ValueString(),
+		Applications: applications,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to revoke secret access", err.Error())
+		return
+	}
+
+	// Save empty list of applications into Terraform state
+	emptyApplicationList, errDiag := types.ListValue(types.StringType, []attr.Value{})
+	resp.Diagnostics.Append(errDiag...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Applications = emptyApplicationList
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+
+	s.trace(fmt.Sprintf("revoke secret access %q", state.SecretId))
+}
+
+// difference returns the difference between two slices.
+func difference(slice1, slice2 interface{}) interface{} {
+	slice1Value := reflect.ValueOf(slice1)
+	slice2Value := reflect.ValueOf(slice2)
+
+	slice1Type := slice1Value.Type()
+	slice2Type := slice2Value.Type()
+
+	if slice1Type.Kind() != reflect.Slice || slice2Type.Kind() != reflect.Slice {
+		panic("Invalid data type. Expected slices.")
+	}
+
+	if slice1Type.Elem() != slice2Type.Elem() {
+		panic("Inconsistent slice types.")
+	}
+
+	set := make(map[interface{}]bool)
+	result := reflect.MakeSlice(slice1Type, 0, slice1Value.Len())
+
+	for i := 0; i < slice1Value.Len(); i++ {
+		set[slice1Value.Index(i).Interface()] = true
+	}
+
+	for i := 0; i < slice2Value.Len(); i++ {
+		if _, found := set[slice2Value.Index(i).Interface()]; found {
+			delete(set, slice2Value.Index(i).Interface())
+		}
```

Comment:
I would call the `result` slice `slice1diff` and get another slice called `slice2diff` (with a `set2`), and do:

```suggestion
		if _, found := set[slice2Value.Index(i).Interface()]; found {
			delete(set, slice2Value.Index(i).Interface())
		} else {
			set2[slice2Value.Index(i).Interface()] = true
		}
```

since the elements you don't find in the `set1` will be the elements that are exclusive to the `slice2`. 

And then we can return both the `slice1diff` and `slice2diff` at the same time, which would i) save us from calling this function twice (to get symmetric difference), therefore ii) save us from performing twice the initializations and validations in the preamble of this function.

Reply:
now there is now simple function implementation